src/plugins/audioplayers/player.c

```diff
@@ -0,0 +1,452 @@
+            gchar *debug;
+
+            gst_message_parse_error(message, &err, &debug);
+            audio_player_on_media_error(data, err, debug);
+            g_error_free(err);
+            g_free(debug);
+            break;
+        }
+        case GST_MESSAGE_STATE_CHANGED: {
+            GstState old_state, new_state;
+
+            gst_message_parse_state_changed(message, &old_state, &new_state, NULL);
+            audio_player_on_media_state_change(data, message->src, &old_state, &new_state);
+            break;
+        }
+        case GST_MESSAGE_EOS:
+            gst_element_set_state(data->playbin, GST_STATE_READY);
+            audio_player_on_playback_ended(data);
+            break;
+        case GST_MESSAGE_DURATION_CHANGED: audio_player_on_duration_update(data); break;
+        case GST_MESSAGE_ASYNC_DONE:
+            if (!data->is_seek_completed) {
+                audio_player_on_seek_completed(data);
+                data->is_seek_completed = true;
+            }
+            break;
+        default:
+            // For more GstMessage types see:
+            // https://gstreamer.freedesktop.org/documentation/gstreamer/gstmessage.html?gi-language=c#enumerations
+            break;
+    }
+
+    // Continue watching for messages
+    return TRUE;
+}
+
+gboolean audio_player_on_refresh(struct audio_player *data) {
+    if (data->playbin->current_state == GST_STATE_PLAYING) {
+        audio_player_on_position_update(data);
+    }
+    return TRUE;
+}
+
+void audio_player_set_playback(struct audio_player *self, int64_t seekTo, double rate) {
+    const GstSeekFlags seek_flags = GST_SEEK_FLAG_FLUSH | GST_SEEK_FLAG_ACCURATE;
+
+    if (!self->is_initialized) {
+        return;
+    }
+    // See:
+    // https://gstreamer.freedesktop.org/documentation/tutorials/basic/playback-speed.html?gi-language=c
+    if (!self->is_seek_completed) {
+        return;
+    }
+    if (rate == 0) {
+        // Do not set rate if it's 0, rather pause.
+        audio_player_pause(self);
+        return;
+    }
+
+    if (self->playback_rate != rate) {
+        self->playback_rate = rate;
+    }
+    self->is_seek_completed = false;
+
+    GstEvent *seek_event;
+    if (rate > 0) {
+        seek_event = gst_event_new_seek(rate, GST_FORMAT_TIME, seek_flags, GST_SEEK_TYPE_SET, seekTo * GST_MSECOND, GST_SEEK_TYPE_NONE, -1);
+    } else {
+        seek_event = gst_event_new_seek(rate, GST_FORMAT_TIME, seek_flags, GST_SEEK_TYPE_SET, 0, GST_SEEK_TYPE_SET, seekTo * GST_MSECOND);
+    }
+    if (!gst_element_send_event(self->playbin, seek_event)) {
+        // FIXME
+        LOG_ERROR("Could not set playback to position %ld and rate %f.", seekTo, rate);
+        self->is_seek_completed = true;
+    }
+}
+void audio_player_on_media_error(struct audio_player *self, GError *error, gchar *debug) {
+    (void) debug;
+    char *error_message = malloc(256 * sizeof(char));
+    snprintf(error_message, 256, "Error: %d; message=%s", error->code, error->message);
```

Comment:
```suggestion
    char *error_message = asprintf(error_message, "Error: %d; message=%s", error->code, error->message);
```

Reply:
Refactored to use static buffer actually